proxy/grpcproxy/lease.go

```
@@ -15,73 +15,326 @@
 }
 
 func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) error {
-	conn := lp.client.ActiveConnection()
+	lp.mu.Lock()
+	select {
+	case <-lp.ctx.Done():
+		lp.mu.Unlock()
+		return lp.ctx.Err()
+	default:
+		lp.wg.Add(1)
+	}
+	lp.mu.Unlock()
+
 	ctx, cancel := context.WithCancel(stream.Context())
-	lc, err := pb.NewLeaseClient(conn).LeaseKeepAlive(ctx)
-	if err != nil {
-		cancel()
-		return err
+	lps := leaseProxyStream{
+		stream:          stream,
+		lessor:          lp.lessor,
+		keepAliveLeases: make(map[int64]*atomicCounter),
+		respc:           make(chan *pb.LeaseKeepAliveResponse),
+		ctx:             ctx,
+		cancel:          cancel,
+		errc:            make(chan error, 2),
 	}
 
+	var leaderc <-chan struct{}
+	if md, ok := metadata.FromContext(stream.Context()); ok {
+		v := md[rpctypes.MetadataRequireLeaderKey]
+		if len(v) > 0 && v[0] == rpctypes.MetadataHasLeader {
+			leaderc = lp.lostLeaderNotify()
+		}
+	}
+	stopc := make(chan struct{}, 3)
 	go func() {
-		// Cancel the context attached to lc to unblock lc.Recv when
-		// this routine returns on error.
-		defer cancel()
-
-		for {
-			// stream.Recv will be unblock when the loop in the parent routine
-			// returns on error.
-			rr, err := stream.Recv()
-			if err != nil {
-				return
-			}
-			err = lc.Send(rr)
-			if err != nil {
-				return
-			}
+		defer func() { stopc <- struct{}{} }()
+		if err := lps.recvLoop(); err != nil {
+			lps.errc <- err
 		}
 	}()
 
-	for {
-		rr, err := lc.Recv()
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		if err := lps.sendLoop(); err != nil {
+			lps.errc <- err
+		}
+	}()
+
+	// tears down LeaseKeepAlive stream if leader goes down or entire leaseProxy is terminated.
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		select {
+		case <-leaderc:
+		case <-ctx.Done():
+		case <-lp.ctx.Done():
+		}
+	}()
+
+	var err error
+	select {
+	case <-stopc:
+		stopc <- struct{}{}
+	case err = <-lps.errc:
```

anyways, should I ignore any errors occurred on stream send() or recv()?